sw/device/silicon_creator/lib/cert/asn1.h

```diff
@@ -0,0 +1,200 @@
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef __ASN1__
+#define __ASN1__
+
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/silicon_creator/lib/error.h"
+
+/** Structure holding the state of the asn1 generator.
+ *
+ * The fields in this structure should be considered
+ * private and not be read or written directly.
+ */
+typedef struct asn1_state {
+  // Buffer containing the output.
+  uint8_t *buffer;
+  // Size of the output buffer.
+  size_t size;
+  // Current offset in the output.
+  size_t offset;
+} asn1_state_t;
+
+/** Start generating an ASN1 stream.
+ *
+ * @param[out] new_state Pointer to the state to be initialized.
+ * @param buffer Pointer to a user-provided buffer.
+ * @param size Size of the user-provided buffer.
+ * @return The result of the operation.
+ */
+rom_error_t asn1_start(asn1_state_t *new_state, uint8_t *buffer, size_t size);
+/** Finish an ASN1 stream and return the size.
+ *
+ * Note: the state will be cleared out after this call.
+ *
+ * @param state Pointer to the state initialized by asn1_start.
+ * @param[out] out_size Pointer to an integer that will be set to the size of
+ * the stream.
+ * @return The result of the operation.
+ */
+rom_error_t asn1_finish(asn1_state_t *state, size_t *out_size);
+
+/** Push a byte into the ASN1 buffer.
+ *
+ * @param state Pointer to the initialized by asn1_start.
+ * @param byte Byte to add to the buffer.
+ * @return The result of the operation.
+ */
+rom_error_t asn1_push_byte(asn1_state_t *state, uint8_t byte);
+
+/** Push bytes into the ASN1 buffer.
+ *
+ * @param state Pointer to the initialized by asn1_start.
+ * @param bytes Pointer to an array of bytes.
+ * @param size Number of bytes in the array.
+ * @return The result of the operation.
+ */
+rom_error_t asn1_push_bytes(asn1_state_t *state, const uint8_t *bytes,
+                            size_t size);
+
+/** Structure holding the information about an unfinished tag sequence.
+ *
+ * The fields in this structure should be considered
+ * private and not be read or written directly.
+ */
+typedef struct asn1_tag {
+  // Pointer to state.
+  asn1_state_t *state;
+  // Offset of the start of the length octets.
+  size_t len_offset;
+  // How many bytes were allocated for the length octets.
+  size_t len_size;
+} asn1_tag_t;
+
+// ASN1 tag classes.
+#define ASN1_CLASS_UNIVERSAL (0 << 6)
+#define ASN1_CLASS_APPLICATION (1 << 6)
+#define ASN1_CLASS_CONTEXT (2 << 6)
+#define ASN1_CLASS_PRIVATE (3 << 6)
+
+// ASN1 tag primitive/constructed bit.
+#define ASN1_FORM_PRIMITIVE (0 << 5)
+#define ASN1_FORM_CONSTRUCTED (1 << 5)
+
+// ASN1 tag number (for universal tags).
+#define ASN1_TAG_BOOLEAN 0x01
+#define ASN1_TAG_INTEGER 0x02
+#define ASN1_TAG_BITSTRING 0x03
+#define ASN1_TAG_OCTET_STRING 0x04
+#define ASN1_TAG_OID 0x06
+#define ASN1_TAG_UTF8_STRING 0x0c
+#define ASN1_TAG_PRINTABLE_STRING 0x13
+#define ASN1_TAG_GENERALIZED_TIME 0x18
+#define ASN1_TAG_SEQUENCE 0x30
+#define ASN1_TAG_SET 0x31
```

Comment:
can we move these to enums? see the style guide https://opentitan.org/book/doc/contributing/style_guides/c_cpp_coding_style.html#constants-and-preprocessor-macros

Reply:
I have made the change. I am afraid it makes the constant names every longer though.